Add Global MIDI Mapping View

fde31 opened the pull request:
This adds a Global MIDI Mapping view. Currently it's read-only apart from clearing the mapping but should be relatively straightforward to make it possible to inline edit the MIDI Channel and CCs.

I've added a button to the left sidebar for now to access this view, reasonable?

Setting to draft for now until editing in place.

Closes #171

x37v commented on src/pages/midimappings.tsx:
```diff
@@ -0,0 +1,138 @@
+import MIDIMappedParameterList from "../components/midi/mappedParameterList";
+import { ParameterRecord } from "../models/parameter";
+import { clearParameterMIDIMappingOnRemote, setParameterMIDIChannelOnRemote, setParameterMIDIControlOnRemote } from "../actions/patchers";
+import { PatcherInstanceRecord } from "../models/instance";
+
+const collator = new Intl.Collator("en-US");
+const parameterComparators: Record<MIDIMappedParameterSortAttr, Record<SortOrder, (a: ParameterRecord, b: ParameterRecord) => number>> = {
+	[MIDIMappedParameterSortAttr.MIDIChannel]: {
```
FYI, there are a few more mappings: https://github.com/Cycling74/rnbo.oscquery.runner?tab=readme-ov-file#midi

x37v commented on src/lib/types.ts:
```diff
@@ -9,6 +9,13 @@ export type AnyJson =
 export interface JsonMap {  [key: string]: AnyJson }
 
+export type ParameterMetaJsonMap = JsonMap & {
+	midi?: {
+		chan?: number;
+		ctrl?: number;
+	}
+};
```
https://github.com/Cycling74/rnbo.oscquery.runner?tab=readme-ov-file#midi

x37v reviewed (commented):
looking good so far, there are a few more types of mappings available that aren't yet visible/implemented that we should get in there

fde31 commented:
Yeah so these other types of mapping were just not really implemented / typed I guess. I've added support for them and moved to a Max-like display / format with some relatively primitive validation in `formatMIDIMappingToDisplay` and `parseMIDIMappingDisplayValue`

x37v commented on src/lib/util.ts:
```diff
@@ -66,3 +67,121 @@ export const formatParamValueForDisplay = (value: number | string) => {
+			}
+			case MIDIMetaMappingType.ControlChange: {
+				const chan = parseInt(match.groups.chan, 10);
+				if (isNaN(chan) || chan < 1 || chan > 16) return false;
+
+				const ctrl = parseInt(match.groups.ctrl, 10);
+				if (isNaN(ctrl) || ctrl < 0 || ctrl > 127) return false;
```
this `if` essentially happens a ton of times.. i'd probably make the whole thing into a func?

```
function parseMIDIByte(val: string, min: number, max: number) {
  let n = parseInt(val, 10);
  if (isNaN(n) || n < min || n > max) {
    return null;
  }
  return n;
}

const ctrl = parseMIDIByte(match.groups.ctrl, 0, 127);
if (ctrl == null) return false;

x37v commented on src/lib/util.ts:
```diff
@@ -66,3 +67,121 @@ export const formatParamValueForDisplay = (value: number | string) => {
+
+			default: {
+				// eslint-disable-next-line @typescript-eslint/no-unused-vars
+				const _exhaustive: never = mappingType;
+				throw new Error(`Unknown MIDIMappingType "${mappingType}"`);
```
this throwing seems okay because it is from a type we've defined here? so it would be an error to not be able to render a known type?

fde31 replied:
this does basically two things - ensure that the switch / case is exhaustive with the `never` type and then in order to have some form of "return" value it throws. But the function is used to parse user input from the MIDI source field to a MIDI mapping.

fde31 commented on src/lib/util.ts:
```diff
@@ -66,3 +67,121 @@ export const formatParamValueForDisplay = (value: number | string) => {
+		default: {
+			// eslint-disable-next-line @typescript-eslint/no-unused-vars
+			const _exhaustive: never = type;
+			throw new Error(`Unknown MIDIMappingType "${type}"`);
```
hmmm. What would we show? `U` followed by a list of `<key>:<value>` mappings?

x37v replied:
maybe something like `return "unknown mapping";` ?

fde31 commented:
k. I've hardened the parsing and display and made it a tad more flexible for _unknown_ mapping types

x37v commented on src/lib/util.ts:
```diff
@@ -105,38 +105,53 @@ const midiMetaRegexp: Record<MIDIMetaMappingType, RegExp> = {
 
-export const parseMIDIMappingDisplayValue = (value: string): false | { type: MIDIMetaMappingType, mapping: MIDIMetaMapping } => {
+const parseMIDIByte = (val: string, min: number, max: number): number | null => {
+	if (val === undefined) return null;
```
tabbing weird here?

fde31 replied:
don't think so. was just temp b4 I fixed the indent, linter doesn't seem to mind either!?!

x37v reviewed (approved):
looks good, just one comment about potential tabbing?
